Mathlib/RingTheory/Nilpotent/GeometricallyReduced.lean

```diff
@@ -0,0 +1,144 @@
+@[mk_iff]
+class IsGeometricallyReduced : Prop where
+  baseChangeReduced : IsReduced (TensorProduct k (AlgebraicClosure k) A)
+
+theorem IsGeometricallyReduced_imp_baseChange_by_closure_Reduced (h : IsGeometricallyReduced k A) :
+    IsReduced (TensorProduct k (AlgebraicClosure k) A) := by
+    exact h.baseChangeReduced
+
+lemma isGeometricallyReduced_of_injective (B : Type) [CommRing B] [Algebra k B] (f : A →ₐ[k] B)
+    (hf : Function.Injective f) [IsGeometricallyReduced k B] : IsGeometricallyReduced k A := by
+  let fK := baseChange  k A B f (AlgebraicClosure k)
+  have hfK : Function.Injective fK := baseChange_by_field_inj k A _ f hf _
+  expose_names
+  rw [isGeometricallyReduced_iff] at *
+  exact isReduced_of_injective fK hfK
+
+
+theorem isGeometricallyReduced_isReduced [IsGeometricallyReduced k A] : IsReduced A := by
+  -- We prove this by providing an injection A →ₐ[k] A ⊗[k] AlgebraicClosure k and then
+  -- applying isReduced_of_injective
+  let f : A →ₐ[k] TensorProduct k (AlgebraicClosure k) A := Algebra.TensorProduct.includeRight
+  expose_names
+  rw [isGeometricallyReduced_iff] at inst_3
+  have hf : Function.Injective f := by
+    apply Algebra.TensorProduct.includeRight_injective
+    exact FaithfulSMul.algebraMap_injective k (AlgebraicClosure k)
+  exact isReduced_of_injective f hf
+
+lemma notReduced_has_nilpotent {R : Type} [Zero R] [Pow R ℕ] (h : ¬IsReduced R) :
+    ∃ x : R, x ≠ 0 ∧ IsNilpotent x := by
+  by_contra h_contra
+  simp only [ne_eq, not_exists, not_and] at h_contra
+  simp only [isReduced_iff, not_forall] at h
+  obtain ⟨x, ⟨hNil, hx⟩⟩ := h
+  tauto
+
+/-- Given a subalgebra C of a k-algebra A, and a k-algebra B, the basechange of C to a subalgebra
+of A ⊗[k] B -/
+def subAlgebraBaseChange (C : Subalgebra k A) : Subalgebra B (TensorProduct k B A) :=
+  AlgHom.range (baseChange k C A C.val B)
+
+lemma FGsubalgebra_baseChange_of_element (x : TensorProduct k A B) :
+    ∃ C : Subalgebra k B , C.FG ∧ x ∈ subAlgebraBaseChange k B A C := by
+  obtain ⟨S, hS⟩ := TensorProduct.exists_finset x
+  let S1 := Set.image (fun j ↦ j.2) S.toSet
+  let C := Algebra.adjoin k S1
+  use C
+  constructor
+  · rw [Subalgebra.fg_def]
+    use S1
+    refine ⟨?_, rfl⟩
+    exact Set.Finite.image (fun j ↦ j.2) (Finset.finite_toSet S)
+  · rw [hS]
+    refine Subalgebra.sum_mem _ ?_
+    intro s hs
+    have hCS : ∀ i ∈ S, i.2 ∈ C := by
+      intro i hi
+      apply SetLike.mem_of_subset
+      · apply Algebra.subset_adjoin
+      · use i
+        exact ⟨hi, rfl⟩
+    use s.1 ⊗ₜ[k] ⟨s.2, hCS s hs⟩
+    rfl
+
+-- If all finitely generated subalgebras of A are geometrically reduced, then A is geometrically
+-- reduced. The result is in https://stacks.math.columbia.edu/tag/030T
+theorem all_FG_geometricallyReduced_isGeometricallyReduced
+    (h : ∀ B : Subalgebra k A, B.FG → IsGeometricallyReduced k B) :
+    IsGeometricallyReduced k A := by
```

Comment:
I haven't looked at this proof yet.

Reply:
this proof has nothing to do with the algebraic closure, or even that k is a field. You should generalise this theorem.